### PR TITLE
[SPARK-39491][YARN] Hadoop 2.7 build fails due to org.apache.hadoop.yarn.api.records.NodeState.DECOMMISSIONING

### PR DESCRIPTION
**Pull request by abhishekd0907**
### What changes were proposed in this pull request?
Spark build fails with hadoop-2 profile, which uses hadoop 2.7 version fails due to:
`
resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:454: value DECOMMISSIONING is not a member of object org.apache.hadoop.yarn.api.records.NodeState
` 
This is because DECOMMISSIONING was only added in hadoop 2.8.

We use string comparison instead of compile-time check and test suite `Test YARN container decommissioning` in `YarnAllocatorSuite` should only run when `VersionUtils.isHadoop3` is true

### Why are the changes needed?

To fix Hadoop 2.7 Build

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Pass GA
Manual test
`run mvn clean install -DskipTests -pl resource-managers/yarn -am -Pyarn -Phadoop-2`

**Before**
```
[ERROR] [Error] /basedir/spark-source/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:454: value DECOMMISSIONING is not a member of object org.apache.hadoop.yarn.api.records.NodeState
[ERROR] one error found
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Spark Project Parent POM 3.4.0-SNAPSHOT:
[INFO] 
[INFO] Spark Project Parent POM ........................... SUCCESS [  3.252 s]
[INFO] Spark Project Tags ................................. SUCCESS [  5.735 s]
[INFO] Spark Project Local DB ............................. SUCCESS [  5.492 s]
[INFO] Spark Project Networking ........................... SUCCESS [  8.251 s]
[INFO] Spark Project Shuffle Streaming Service ............ SUCCESS [  6.334 s]
[INFO] Spark Project Unsafe ............................... SUCCESS [ 15.326 s]
[INFO] Spark Project Launcher ............................. SUCCESS [  4.905 s]
[INFO] Spark Project Core ................................. SUCCESS [02:07 min]
[INFO] Spark Project YARN Shuffle Service ................. SUCCESS [ 17.382 s]
[INFO] Spark Project YARN ................................. FAILURE [  7.718 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  03:22 min
[INFO] Finished at: 2022-06-20T11:57:54+08:00
[INFO] ------------------------------------------------------------------------
```
**After**
```
[INFO] Reactor Summary for Spark Project Parent POM 3.4.0-SNAPSHOT:
[INFO] 
[INFO] Spark Project Parent POM ........................... SUCCESS [  5.451 s]
[INFO] Spark Project Tags ................................. SUCCESS [  5.739 s]
[INFO] Spark Project Local DB ............................. SUCCESS [  5.908 s]
[INFO] Spark Project Networking ........................... SUCCESS [  8.310 s]
[INFO] Spark Project Shuffle Streaming Service ............ SUCCESS [  5.857 s]
[INFO] Spark Project Unsafe ............................... SUCCESS [  8.439 s]
[INFO] Spark Project Launcher ............................. SUCCESS [  4.795 s]
[INFO] Spark Project Core ................................. SUCCESS [02:36 min]
[INFO] Spark Project YARN Shuffle Service ................. SUCCESS [ 15.044 s]
[INFO] Spark Project YARN ................................. SUCCESS [ 32.517 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  04:09 min
[INFO] Finished at: 2022-06-20T13:10:04+08:00
[INFO] ------------------------------------------------------------------------
```
Run `run mvn clean install -pl resource-managers/yarn -Pyarn -Phadoop-2 -Dtest=none -DwildcardSuites=org.apache.spark.deploy.yarn.YarnAllocatorSuite`

```
- Test YARN container decommissioning !!! CANCELED !!!
  org.apache.spark.util.VersionUtils.isHadoop3 was false (YarnAllocatorSuite.scala:749)
Run completed in 2 seconds, 140 milliseconds.
Total number of tests run: 16
Suites: completed 2, aborted 0
Tests: succeeded 16, failed 0, canceled 6, ignored 0, pending 0
All tests passed.
```

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Hm, but we lately release it with Hadoop 2.7 build though. which command did you use to build?

**Comment by abhishekd0907** [comment]
> Hm, but we lately release it with Hadoop 2.7 build though. which command did you use to build?

@tgravescs since you filed the JIRA, can you please share the command you were using?

**Comment by abhishekd0907** [comment]
@tgravescs @mridulm @attilapiros @HyukjinKwon @LuciferYang 
please review,

**Comment by LuciferYang** [comment]
[SPARK-30835](https://issues.apache.org/jira/browse/SPARK-30835) already RESOLVED

**Comment by HyukjinKwon** [comment]
@abhishekd0907 how did you test this patch?

**Comment by LuciferYang** [comment]
@HyukjinKwon I test manually as follows:

```
mvn clean install -DskipTests -pl resource-managers/yarn -am -Pyarn -Phadoop-2
```

```
[INFO] Compiling 25 Scala sources to /basedir/spark/resource-managers/yarn/target/scala-2.12/classes ...
[ERROR] [Error] /basedir/spark/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:454: value DECOMMISSIONING is not a member of object org.apache.hadoop.yarn.api.records.NodeState
[ERROR] one error found
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Spark Project Parent POM 3.4.0-SNAPSHOT:
[INFO] 
[INFO] Spark Project Parent POM ........................... SUCCESS [  3.637 s]
[INFO] Spark Project Tags ................................. SUCCESS [ 10.738 s]
[INFO] Spark Project Local DB ............................. SUCCESS [  5.492 s]
[INFO] Spark Project Networking ........................... SUCCESS [  8.388 s]
[INFO] Spark Project Shuffle Streaming Service ............ SUCCESS [  5.692 s]
[INFO] Spark Project Unsafe ............................... SUCCESS [  7.937 s]
[INFO] Spark Project Launcher ............................. SUCCESS [  4.376 s]
[INFO] Spark Project Core ................................. SUCCESS [02:56 min]
[INFO] Spark Project YARN Shuffle Service ................. SUCCESS [ 24.754 s]
[INFO] Spark Project YARN ................................. FAILURE [ 12.485 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  04:20 min
[INFO] Finished at: 2022-06-18T20:18:16+08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal net.alchim31.maven:scala-maven-plugin:4.3.0:compile (scala-compile-first) on project spark-yarn_2.12: Execution scala-compile-first of goal net.alchim31.maven:scala-maven-plugin:4.3.0:compile failed.: CompileFailed -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :spark-yarn_2.12
```

**Comment by LuciferYang** [comment]
Let me check with this pr later

**Comment by LuciferYang** [comment]
> @HyukjinKwon I test manually as follows without this pr:
> 
> ```
> mvn clean install -DskipTests -pl resource-managers/yarn -am -Pyarn -Phadoop-2
> ```
> 
> ```
> [INFO] Compiling 25 Scala sources to /basedir/spark/resource-managers/yarn/target/scala-2.12/classes ...
> [ERROR] [Error] /basedir/spark/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:454: value DECOMMISSIONING is not a member of object org.apache.hadoop.yarn.api.records.NodeState
> [ERROR] one error found
> [INFO] ------------------------------------------------------------------------
> [INFO] Reactor Summary for Spark Project Parent POM 3.4.0-SNAPSHOT:
> [INFO] 
> [INFO] Spark Project Parent POM ........................... SUCCESS [  3.637 s]
> [INFO] Spark Project Tags ................................. SUCCESS [ 10.738 s]
> [INFO] Spark Project Local DB ............................. SUCCESS [  5.492 s]
> [INFO] Spark Project Networking ........................... SUCCESS [  8.388 s]
> [INFO] Spark Project Shuffle Streaming Service ............ SUCCESS [  5.692 s]
> [INFO] Spark Project Unsafe ............................... SUCCESS [  7.937 s]
> [INFO] Spark Project Launcher ............................. SUCCESS [  4.376 s]
> [INFO] Spark Project Core ................................. SUCCESS [02:56 min]
> [INFO] Spark Project YARN Shuffle Service ................. SUCCESS [ 24.754 s]
> [INFO] Spark Project YARN ................................. FAILURE [ 12.485 s]
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD FAILURE
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time:  04:20 min
> [INFO] Finished at: 2022-06-18T20:18:16+08:00
> [INFO] ------------------------------------------------------------------------
> [ERROR] Failed to execute goal net.alchim31.maven:scala-maven-plugin:4.3.0:compile (scala-compile-first) on project spark-yarn_2.12: Execution scala-compile-first of goal net.alchim31.maven:scala-maven-plugin:4.3.0:compile failed.: CompileFailed -> [Help 1]
> [ERROR] 
> [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
> [ERROR] Re-run Maven using the -X switch to enable full debug logging.
> [ERROR] 
> [ERROR] For more information about the errors and possible solutions, please read the following articles:
> [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
> [ERROR] 
> [ERROR] After correcting the problems, you can resume the build with the command
> [ERROR]   mvn <args> -rf :spark-yarn_2.12
> ```

Fortunately, v3.3 BUILD SUCCESS

**Comment by LuciferYang** [comment]
> @HyukjinKwon I test manually as follows without this pr:
> 
> ```
> mvn clean install -DskipTests -pl resource-managers/yarn -am -Pyarn -Phadoop-2
> ```
> 
> ```
> [INFO] Compiling 25 Scala sources to /basedir/spark/resource-managers/yarn/target/scala-2.12/classes ...
> [ERROR] [Error] /basedir/spark/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:454: value DECOMMISSIONING is not a member of object org.apache.hadoop.yarn.api.records.NodeState
> [ERROR] one error found
> [INFO] ------------------------------------------------------------------------
> [INFO] Reactor Summary for Spark Project Parent POM 3.4.0-SNAPSHOT:
> [INFO] 
> [INFO] Spark Project Parent POM ........................... SUCCESS [  3.637 s]
> [INFO] Spark Project Tags ................................. SUCCESS [ 10.738 s]
> [INFO] Spark Project Local DB ............................. SUCCESS [  5.492 s]
> [INFO] Spark Project Networking ........................... SUCCESS [  8.388 s]
> [INFO] Spark Project Shuffle Streaming Service ............ SUCCESS [  5.692 s]
> [INFO] Spark Project Unsafe ............................... SUCCESS [  7.937 s]
> [INFO] Spark Project Launcher ............................. SUCCESS [  4.376 s]
> [INFO] Spark Project Core ................................. SUCCESS [02:56 min]
> [INFO] Spark Project YARN Shuffle Service ................. SUCCESS [ 24.754 s]
> [INFO] Spark Project YARN ................................. FAILURE [ 12.485 s]
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD FAILURE
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time:  04:20 min
> [INFO] Finished at: 2022-06-18T20:18:16+08:00
> [INFO] ------------------------------------------------------------------------
> [ERROR] Failed to execute goal net.alchim31.maven:scala-maven-plugin:4.3.0:compile (scala-compile-first) on project spark-yarn_2.12: Execution scala-compile-first of goal net.alchim31.maven:scala-maven-plugin:4.3.0:compile failed.: CompileFailed -> [Help 1]
> [ERROR] 
> [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
> [ERROR] Re-run Maven using the -X switch to enable full debug logging.
> [ERROR] 
> [ERROR] For more information about the errors and possible solutions, please read the following articles:
> [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
> [ERROR] 
> [ERROR] After correcting the problems, you can resume the build with the command
> [ERROR]   mvn <args> -rf :spark-yarn_2.12
> ```

with this pr build success

```
[INFO] Reactor Summary for Spark Project Parent POM 3.4.0-SNAPSHOT:
[INFO] 
[INFO] Spark Project Parent POM ........................... SUCCESS [  3.747 s]
[INFO] Spark Project Tags ................................. SUCCESS [ 13.542 s]
[INFO] Spark Project Local DB ............................. SUCCESS [ 13.811 s]
[INFO] Spark Project Networking ........................... SUCCESS [ 17.388 s]
[INFO] Spark Project Shuffle Streaming Service ............ SUCCESS [  9.835 s]
[INFO] Spark Project Unsafe ............................... SUCCESS [ 11.792 s]
[INFO] Spark Project Launcher ............................. SUCCESS [  9.364 s]
[INFO] Spark Project Core ................................. SUCCESS [03:02 min]
[INFO] Spark Project YARN Shuffle Service ................. SUCCESS [ 30.487 s]
[INFO] Spark Project YARN ................................. SUCCESS [ 33.975 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  05:27 min
[INFO] Finished at: 2022-06-18T21:32:30+08:00
[INFO] ------------------------------------------------------------------------

```

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala [hunk 1]
```diff
@@ -462,6 +463,21 @@ private[yarn] class YarnAllocator(
 
   private def getHostAddress(nodeReport: NodeReport): String = nodeReport.getNodeId.getHost
 
+  /**
+   * Checks if node is in DECOMMISSIONING STATE
+   *
+   * Visible for testing.
+   */
+  def isDecommissioningNode(node: NodeReport): Boolean = {
```
Comment (LuciferYang) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:471]: Can we change this to `private` and use `invokePrivate` in test suite?

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala [hunk 2]
```diff
@@ -1004,3 +1020,35 @@ private object YarnAllocator {
...
+  // Helper method to get NodeState of the Yarn.
+  def getYarnNodeState(state: YarnNodeState): NodeState.Value = {
+    // In hadoop-2.7 there is no support for node state DECOMMISSIONING
+    // In Hadoop-2.8, hadoop3.1 and later version of spark there is a support
+    // to node state DECOMMISSIONING.
+    // Inorder to build the spark using hadoop2 and hadoop3,
+    // using string comparison for the YARN node state DECOMMISSIONING here and
+    // for other states we are matching the YarnNodeState and assigning
+    // the node state at spark end
+    if (state.toString.equals(NodeState.DECOMMISSIONING.toString)) {
```
Comment (LuciferYang) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:1041]: should we change to
```scala
state match {
      case YarnNodeState.RUNNING => NodeState.RUNNING
      case YarnNodeState.DECOMMISSIONED => NodeState.DECOMMISSIONED
      case YarnNodeState.LOST => NodeState.LOST
      case YarnNodeState.UNHEALTHY => NodeState.LOST
      case other if other.toString.equals(NodeState.DECOMMISSIONING.toString) =>
        NodeState.DECOMMISSIONING
      case _ => NodeState.OTHER
    }
```

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala [hunk 3]
```diff
@@ -462,6 +463,21 @@ private[yarn] class YarnAllocator(
...
+   * Visible for testing.
+   */
+  def isDecommissioningNode(node: NodeReport): Boolean = {
+    // In hadoop-2.7 there is no support for node state DECOMMISSIONING
+    // In Hadoop-2.8, hadoop3.1 and later version of spark there is a support
+    // to node state DECOMMISSIONING.
+    // Inorder to build the spark using hadoop2 and hadoop3, not
+    // using YarnNodeState for the node state DECOMMISSIONING here and
+    // instead comparing Node states on spark end
+    NodeState.getYarnNodeState(node.getNodeState) == NodeState.DECOMMISSIONING
```
Comment (LuciferYang) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:478]: hmm... Is it really necessary to define `NodeState`?

Is
```
NodeState.getYarnNodeState(node.getNodeState) == NodeState.DECOMMISSIONING
```
equivalent to 

```
node.getNodeState.toString.equals("DECOMMISSIONING")
```
?

Comment (LuciferYang) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:478]: It seems that just change line 454 from 

```
node.getNodeState == NodeState.DECOMMISSIONING
```

to

```
node.getNodeState.toString.equals("DECOMMISSIONING")
``` 

and add some comments is enough

**Comment by LuciferYang** [comment]
Any updates of this PR?

**Comment by abhishekd0907** [comment]
> Any updates of this PR?

@LuciferYang I have updated the PR as per your comments and your related PR #36917 . Sorry for the delay.

**Comment by HyukjinKwon** [comment]
Let me merge https://github.com/apache/spark/pull/36917 with adding @abhishekd0907 as a co-author since that approach was proposed in that Pr first.

**Pull request closed without merging**